File: forms/patient-intake/front-end-patient-form-with-svelte/src/lib/engine/intake-grader.ts
import type { AssessmentData, RiskLevel, FiredRule } from './types';
import { intakeRules } from './intake-rules';

/**
 * Pure function: evaluates all intake rules against patient data.
 * Returns the maximum risk level among all fired rules (worst complexity),
 * defaulting to Low for healthy patients with no fired rules.
 */
export function calculateRiskLevel(data: AssessmentData): {
	riskLevel: RiskLevel;
	firedRules: FiredRule[];
} {
	const firedRules: FiredRule[] = [];

	for (const rule of intakeRules) {
		try {
			if (rule.evaluate(data)) {
				firedRules.push({
					id: rule.id,
					category: rule.category,
					description: rule.description,
					riskLevel: rule.riskLevel
				});
			}
		} catch (e) {
			// Rule evaluation failed - log for debugging but continue grading
			console.warn(`Intake rule ${rule.id} evaluation failed:`, e);
		}
	}

	const riskOrder: Record<RiskLevel, number> = { low: 0, medium: 1, high: 2 };

	const riskLevel: RiskLevel =
		firedRules.length === 0
			? 'low'
			: firedRules.reduce<RiskLevel>((max, r) => {
					return riskOrder[r.riskLevel] > riskOrder[max] ? r.riskLevel : max;
				}, 'low');

	return { riskLevel, firedRules };
}

File: forms/patient-intake/full-stack-with-rust-axum-loco-tera-htmx-alpine/src/engine/intake_grader.rs
use super::intake_rules::all_rules;
use super::types::{AssessmentData, FiredRule, RiskLevel};

/// Pure function: evaluates all intake rules against patient data.
/// Returns the maximum risk level among all fired rules (worst complexity),
/// defaulting to "low" for healthy patients with no fired rules.
pub fn calculate_risk_level(data: &AssessmentData) -> (RiskLevel, Vec<FiredRule>) {
    let mut fired_rules = Vec::new();

    for rule in all_rules() {
        if (rule.evaluate)(data) {
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                risk_level: rule.risk_level.to_string(),
            });
        }
    }

    let risk_order = |level: &str| -> u8 {
        match level {
            "low" => 0,
            "medium" => 1,
            "high" => 2,
            _ => 0,
        }
    };

    let risk_level = if fired_rules.is_empty() {
        "low".to_string()
    } else {
        fired_rules
            .iter()
            .map(|r| r.risk_level.as_str())
            .max_by_key(|l| risk_order(l))
            .unwrap_or("low")
            .to_string()
    };

    (risk_level, fired_rules)
}
